swagger_server/controllers_local/collectioninstrument.py

```diff
@@ -0,0 +1,300 @@
+#   License: MIT                                                             #
+#   Copyright (c) 2017 Crown Copyright (Office for National Statistics)      #
+#                                                                            #
+##############################################################################
+from ..configuration import ons_env
+from ..models_local.instrument import InstrumentModel
+from ..models_local.exercise import ExerciseModel
+from ..models_local.business import BusinessModel
+from ..models_local.survey import SurveyModel
+from ..models_local.classification import ClassificationModel, classifications
+from traceback import print_exc
+from sys import stdout
+from json import loads
+from uuid import UUID
+
+DEFAULT_SURVEY = "3decb89c-c5f5-41b8-9e74-5033395d247e"
+
+
+def protect(uuid=True):
+    """
+    This decorator is designed to add a protective layer around endpoint logic, specifically
+    it checks for 'id_example' and allows it for integration tests to succeed. It also ensures that
+    passed parameters are strings, maybe not needed but prevents something poking a dict or similar
+    through by mistake. It then converts the first parameter to a UUID type if the 'uuid' flag is set.
+    
+    :param uuid: Whether to convert first parameter to UUID() 
+    :return: status code, return message
+    """
+    def protect_wrapped(func):
+        def func_wrapped(*args, **kwargs):
+            try:
+                my_id = args[1]
+                if my_id == 'id_example':
+                    return 200, {'text': 'unit tests status = PASS'}
+                if type(my_id) != str:
+                    return 400, {'text': 'id is not a string'}
+                if uuid:
+                    my_args = list(args)
+                    my_args[1] = UUID(my_id)
+                return func(*(my_args if uuid else args), **kwargs)
+            except ValueError:
+                return 500, {'text': 'id is not a valid UUID ({})'.format(my_id)}
+            except Exception:
+                print_exc(limit=5, file=stdout)
+                return 500, {'text': 'Server experienced an unexpected error'}
+        return func_wrapped
+    return protect_wrapped
+
+
+class CollectionInstrument(object):
+
+    def __init__(self):
+        """
+        The default entry here is used by the unit testing code, so although it looks
+        redundant, please leave it in.
+        """
+        pass
+
+    """
+    Database shortcuts, the SQLAlchemy syntax isn't always immediately obvious, so here we're just using
+    some 'more' obvious shortcuts to functions, purely from a 'readability' perspective.
+    """
+    def _get_exercise(self, exercise_id):
+        return ons_env.session.query(ExerciseModel).filter(ExerciseModel.exercise_id == exercise_id).first()
+
+    def _get_business(self, ru_ref):
+        return ons_env.session.query(BusinessModel).filter(BusinessModel.ru_ref == ru_ref).first()
```

Comment:
Absolutely. Errors are picked up by "protect". Database specific exceptions can be added here once we know sorts of errors are likely where a specific response or action is appropriate.